Openstack LB improvements

anguslees opened the pull request:
In particular, makes EnsureTCPLoadBalancerDeleted idempotent, and fixes an existing bug that leaked monitors on LB delete.

anguslees committed:
Catch 404 and return exists=false from GetTCPLoadBalancer

Previouly getVipByName treated 404 like any other unexpected error
return and passed it up the chain.  This caused the "if ErrNotFound then
exists=false" logic in GetTCPLoadBalancer to never fire.

This change teaches getVipByName to return ErrNotFound on a 404 server
response.

anguslees committed:
Actually delete LBaaS monitors after disassociating

Partially addresses issue #8352

anguslees committed:
Make EnsureTCPLoadBalancerDeleted idempotent

This change allows EnsureTCPLoadBalancerDeleted to be called repeatedly
to reattempt deleting objects that may have failed on a previous run.

Specifically, if the VIP is already deleted, then an attempt is made to
lookup the pool by name.  Returns success when both the VIP and pool are
not found.

Fixes #8352

erictune commented:
OK to test
 On Jun 2, 2015 12:34 AM, "Kubernetes Bot" <notifications@github.com> wrote:

> Can one of the admins verify that this patch is reasonable to test? (reply
> "ok to test", or if you trust the user, reply "add to whitelist")
>
> If this message is too spammy, please complain to ixdy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9105#issuecomment-107840186>
> .
>

erictune commented:
Add to whitelist
On Jun 2, 2015 5:25 AM, "Eric Tune" <etune@google.com> wrote:

> OK to test
>  On Jun 2, 2015 12:34 AM, "Kubernetes Bot" <notifications@github.com>
> wrote:
>
>> Can one of the admins verify that this patch is reasonable to test?
>> (reply "ok to test", or if you trust the user, reply "add to whitelist")
>>
>> If this message is too spammy, please complain to ixdy.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9105#issuecomment-107840186>
>> .
>>
>

bgrant0607 commented:
cc @justinsb @thockin

a-robinson commented on pkg/cloudprovider/openstack/openstack.go:
```diff
@@ -651,45 +696,53 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
-			// TODO(#8352): Delete the monitor, don't just disassociate it.
-			pools.DisassociateMonitor(lb.network, pool.ID, monId)
+			_, err = pools.DisassociateMonitor(lb.network, pool.ID, monId).Extract()
+			if err != nil {
```
Is it possible for this to be a not-found error?

anguslees replied:
They can't be deleted while associated to a pool, but it is possible that another routine much like this one is racing with us and has already disassociated/deleted various portions of the lbaas setup.
I figured at this point exiting out and restarting the entire delete process was a fine response.

a-robinson commented on pkg/cloudprovider/openstack/openstack.go:
```diff
@@ -651,45 +696,53 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
+			}
+
+			err = monitors.Delete(lb.network, monId).ExtractErr()
+			if err != nil {
```
Probably not needed since the retry would succeed, but if some parallel process were also messing with the resources then this could conceivably be a not-found error.

anguslees replied:
Yep, again I figured it was reasonable to restart the entire delete cycle in this case.

Note also that if the Disassociate succeeds but the Delete fails then we will leak the monitor.  We can't set custom names on monitors so the only way I could track them would be to record their server-assigned ID somewhere.

anguslees replied:
I just added the isNotFound check here too.   I didn't have a strong reason for behaving differently here, and being consistent at least makes future reviewers not wonder why it was ok here but not elsewhere ;)

a-robinson commented on pkg/cloudprovider/openstack/openstack.go:
```diff
@@ -443,6 +443,48 @@ func (os *OpenStack) TCPLoadBalancer() (cloudprovider.TCPLoadBalancer, bool) {
+		return true, nil
+	})
+	if err != nil {
+		if e, ok := err.(*gophercloud.UnexpectedResponseCodeError); ok {
```
Shouldn't this be using isNotFound?

anguslees replied:
Oh indeed!  I distinctly recall making that change already, so it must have been lost in some git remangling.  Retyped :/

a-robinson commented on pkg/cloudprovider/openstack/openstack.go:
```diff
@@ -443,6 +443,48 @@ func (os *OpenStack) TCPLoadBalancer() (cloudprovider.TCPLoadBalancer, bool) {
+	poolList := make([]pools.Pool, 0, 1)
+
+	err := pager.EachPage(func(page pagination.Page) (bool, error) {
+		v, err := pools.ExtractPools(page)
```
it'd probably make sense to change from v to p for the variable name

anguslees replied:
yep, changed.

a-robinson commented:
Thanks for the improvements, @anguslees! Can you also squash the gofmt commit when you get a chance?

This fixes #8352

justinsb commented:
I don't suppose there is a comparable change coming to make the Create logic idempotent :-) ?

Not related to this PR, but is there a reason the externalIP must be specified for OpenStack LBs (vs just creating one on-demand)?  I think that API functionality is at least deprecated, and we'd like to kill it altogether.

a-robinson commented:
LGTM

bgrant0607 commented:
@anguslees Any chance you could make the CLA bot happy?

quinton-hoole commented:
@anguslees Have you signed the CLA?

quinton-hoole commented:
@anguslees We need an SLA from your company before merging.

anguslees commented:
@quinton-hoole,@bgrant0607: I'm doing this work on Rackspace company time, and Rackspace has signed the Google Corporate CLA.  None of the helpful googlebot links let me explore the situation further, but I suspect googlebot only knows about my @rackspace.com address, which I don't use on github, etc.  If there was a way to get googlebot to understand this, that would be great.  Currently I repeat some version of this explanation on every Kubernetes commit.

anguslees commented:
@justinsb: huh, the externalIP is meant to be optional...

Oh, from eyeballing the code, I'm guessing the externalIP ends up as 0.0.0.0 or something terrible.  I'll add an explicit check for externalIP.IsUnspecified() ...

anguslees committed:
Ignore "unspecified" externalIP during LB create

Previously we always passed `Address: externalIP.String()` while
creating a loadbalancer VIP.  This passed "0.0.0.0" when externalIP was
unspecified, effectively making it mandatory to specify an externalIP.

This change correctly leaves `Address` unspecified when externalIP is
unspecified (has a zero value).

(Thanks to @justinsb for the report)

anguslees commented:
@justinsb: I think I've fixed the externalIP issue.

Re idempotent Create - I'm waiting for https://github.com/GoogleCloudPlatform/kubernetes/pull/8366 to reach a conclusion first ;)  There's actually not a lot I can change on an OpenStack LBaaS VIP without deleting+recreating anyway (only really cosmetic name/description and pool members).

quinton-hoole commented:
Hi Justin

I've manually verified your CLA for you.  In future, to avoid the hassle you should:

1. Add your rackspace.com email address to your github account
2. Make sure that Rackspace have set up an appropriate Google Group and made you a member of it.  Rackspace should have received instructions for this along with the CLA they signed.

Thanks
Q

tonanhngo commented:
Hi @anguslees,
    I am working on OpenStack Magnum and I am trying to get createExternalLoadBalancer to work but have not been successful on a Fedora Atomic image.  I have configured kube-apiserver and kube-controller-manager with the option --cloud_config and --cloud_provider, but they either crash or would not work properly with Neutron. Is there some instruction on how to set up a known working environment?
Thanks,
Ton Ngo
